Remove doubled exception message from nunit

CaptnCodr opened the pull request:
As mentioned here: https://github.com/fsprojects/FsUnit/issues/166#issuecomment-661297280

CaptnCodr commented:
I'm not quite lucky about the `Environment.NewLine Environment.NewLine`.
@sergey-tihon Do you have an idea to make it prettier?

sergey-tihon commented:
I am not sure about...

```fsharp
type InitMsgUtils() =
     inherit FSharpCustomMessageFormatter()
```


I (somehow) missed this line in docs for years and I think that I am not along here.
It is quite not intuitive ... it would be much easier to just reference FsUnit and use operators.

Can we remove `FSharpCustomMessageFormatter` from docs and continue use your implementation?
will it fix double error message?

CaptnCodr commented:
Yes, I agree.

I will overthink this approach to have a neat implementation that nobody have to add this piece of code to their implementation. It's not quite intuitive.

Now: When we remove `FSharpCustomMessageFormatter` then assertion messages will print "ugly" in some cases. No doubled assertion output.

Goal: Removing `FSharpCustomMessageFormatter` and pretty printed messages.

I go for it.

CaptnCodr commented:
It now works, @sergey-tihon. 

I removed `type InitMsgUtils() = inherit FSharpCustomMessageFormatter()` from tests and added an instantiation to the `shouldEqual` and `shouldNotEqual` functions.

- The user **not** have to add `type InitMsgUtils() = inherit FSharpCustomMessageFormatter()` to their code.
- It pretty prints the special types and no double assertion messages.
- It prints the native NUnit messages (`"Expected: ... But was: ..."`).

sergey-tihon commented on src/FsUnit.NUnit/FsUnitTyped.fs:
```diff
@@ -5,15 +5,17 @@ open NUnit.Framework
     [<DebuggerStepThrough>]
     let shouldEqual (expected: 'a) (actual: 'a) =
-        Assert.That(actual, FsUnit.Equality.IsEqualTo(expected), sprintf "Expected: %A\nActual: %A" expected actual)
+        FsUnit.TopLevelOperators.FSharpCustomMessageFormatter() |> ignore
```
I believe that we should do it only once, but not on every call...
Right now it looks like we call `TestContext.AddFormatter(ValueFormatterFactory(fun _ -> ValueFormatter(sprintf "%A")))` on every assertion.

CaptnCodr replied:
Do you have an idea how to implicitly call it?

sergey-tihon replied:
No, we have to do it explicitly, bug we can do it using `lazy` and ensure that only one `ValueFormatter` is registered in `TestContext`

sergey-tihon replied:
If we want to support `FSharpCustomMessageFormatter` in existing solution, then we should add static lazy coda that register formatter and force it from FSharpCustomMessageFormatter constructor, `shouldEqual` and `shouldNotEqual`

https://github.com/fsprojects/FsUnit/blob/67dcc91f8b5cf60b1b80fb3a63621136f9df9344/src/FsUnit.NUnit/FsUnit.fs#L13

CaptnCodr replied:
Unfortunately I'm not able to do it the `lazy` way or I'm doing it wrong. :(
Do you have a reference how to do it properly?

CaptnCodr replied:
We may have to add this formatter in every context in NUnit.
I tested: `[ Ok "ok" ] |> should be (supersetOf [ Ok "no" ])` 
It is printed ugly aswell. When I add `FSharpCustomMessageFormatter |> ignore` to `FsUnit.should` then it's pretty.

CaptnCodr replied:
I don't really know how we go on now.

sergey-tihon commented:
>: We could remove FSharpCustomMessageFormatter from the docs but it's still supported from our side though it's not necessary anymore. What do you think?

Agree, let remove it from here 
https://github.com/fsprojects/FsUnit/blob/master/docs/content/NUnit.fsx#L25-L33

sergey-tihon commented:
https://docs.nunit.org/articles/nunit/writing-tests/attributes/setupfixture.html

> The OneTimeSetUp method in a SetUpFixture is executed once before any of the fixtures contained in its namespace.

`FSharpCustomMessageFormatter` affects only tests from the same namespace 😢

> A SetUpFixture outside of any namespace provides SetUp and TearDown for the entire assembly.

It should be defined in `global` namespace to affect whole assembly 😢


Looks like there is no better way to add custom formatter to NUnit rather than `SetUpFixture`  in `global` namespace

CaptnCodr commented:
Do we another chance leave it like it is? or do you have another plan?

> FSharpCustomMessageFormatter affects only tests from the same namespace 😢

And `FSharpCustomMessageFormatter` have to be initialized - in our code or in the user's code.

sergey-tihon commented:
> Do we another chance leave it like it is? or do you have another plan?
I think that we can merge this PR as it is and improve FsUnit docs with information that `FSharpCustomMessageFormatter` is required, why and how to use it.

> And FSharpCustomMessageFormatter have to be initialized to work properly - in our code or in the user's code.
We cannot  cover all cases from out code =( because user can directly can NUnit functions without FsUnit operators.

CaptnCodr commented:
I think there are two ways to use it:
-> Implement `type InitMsgUtils() = inherit FSharpCustomMessageFormatter()` as before in the assembly where is it required.
or
-> Initialize `FSharpCustomMessageFormatter()` where this is needed e.g. when the `Result`-Type is to be used.

We could add the docs changes in this PR, couldn't we?!

sergey-tihon commented:
> We could add the docs changes in this PR, couldn't we?!
yes, sure

CaptnCodr commented:
I will test some things later this day.

CaptnCodr commented:
@sergey-tihon Do we still need the init.fs? 
Do we have tests that they have "ugly" print? 😄

CaptnCodr commented:
> Do we have tests that they have "ugly" print?

The best way to test this is to implement tests for each operator. 😅

sergey-tihon commented:
> Do we still need the init.fs?

Yes, at least your new `Assert.Throws<AssertionException>` tests

> The best way to test this is to implement tests for each operator. 😅

Yes, awesome idea ;)

> ... from [<SetUp>].

Inside each test fixture? can you show sample?

CaptnCodr commented:
> Yes, at least your new Assert.Throws<AssertionException> tests

We only need this when we have the special types.

> Yes, awesome idea ;)

I will start a marathon next week to implement this. But first let us merge this PR. ;)

> Inside each test fixture? can you show sample?

When it's needed then the user can add it like here:
https://github.com/fsprojects/FsUnit/pull/167/commits/e0872ebc8bea562074b9d5f9b7255eaf0acf7f83#diff-769022280a29bb055c8335bafad02265R10-R12

I adjusted the doc to that effect:
https://github.com/fsprojects/FsUnit/pull/167/commits/e0872ebc8bea562074b9d5f9b7255eaf0acf7f83#diff-a60cf0d77453ee3950367857e2bf74d4R35-R48

CaptnCodr commented:
> Do we have tests that they have "ugly" print?

Every operator in `FsUnit.NUnit` is affected to have ugly printing but putting `type InitMsgUtils() = inherit FSharpCustomMessageFormatter()` to `init.fs` in `namespace global` is the best way to have this for **all** tests.

For single tests the user can add `[<SetUp>] member __.setup () = FSharpCustomMessageFormatter() |> ignore` to the `TestFixture` or `FSharpCustomMessageFormatter() |> ignore` in _one_ test.

I see the effort that is too high to the test all operators. We have no chance to take the work from the user.

We could try to resolve this with https://github.com/nunit/nunit/ to have a effortless integration. 
@sergey-tihon Do you want to open an issue for that?

I could remove `[<SetUp>] member __.setup () = FSharpCustomMessageFormatter() |> ignore` from this PR to stay clean. I'm a bit disappointed about this behaviour and that we can not fix this on our side.

CaptnCodr commented:
When somebody is testing his/her fsharp project within a C#-Project an `Result`-Type it's appearing there too like this:
![image](https://user-images.githubusercontent.com/4574031/88403260-9c729980-cdcc-11ea-8439-eeb968bbd834.png)
This is from a C# NUnit project with folowing code:
```
var ok = FSharpResult<string, string>.NewOk("ok");
var nook = FSharpResult<string, string>.NewOk("not ok");
Assert.AreEqual(ok, nook);
```

CaptnCodr commented:
@sergey-tihon or I implement a special operator that handles the `Result`-Type (`shouldResult` or `equalResult`) on our side. 🤣
What do you think? 

That could look like:
`Ok "ok" |> shouldResult Ok "ok"` or
`Ok "ok" |> should equalResult (Ok "ok")`

(Sorry, for the mass of comments and mentions.)

sergey-tihon commented:
I've removed setup methods from our tests, because I do not  understand why we need them together with init.fs

Why do we need `shouldEqualResult`? Does not it covered by `FSharpCustomMessageFormatter`? We cannot operator for every F# specific type  (Result from FSharp.Core, Result from Chessie and etc)

CaptnCodr commented:
Yes, I see. 
We cannot implement everything for every purpose.
